src/with-urql-client.tsx

```diff
@@ -1,106 +1,75 @@
-import { AppContext } from 'next/app';
+import { NextPage, NextPageContext } from 'next';
+import NextApp, { AppContext } from 'next/app';
 import ssrPrepass from 'react-ssr-prepass';
-import {
-  Provider,
-  Client,
-  ClientOptions,
-  dedupExchange,
-  cacheExchange,
-  fetchExchange,
-  Exchange,
-} from 'urql';
-import { SSRExchange, SSRData } from 'urql/dist/types/exchanges/ssr';
+import { Provider, dedupExchange, cacheExchange, fetchExchange } from 'urql';
 
 import { initUrqlClient } from './init-urql-client';
-
-export type NextUrqlClientOptions = Omit<
-  ClientOptions,
-  'exchanges' | 'suspense'
->;
-
-interface WithUrqlClient {
-  urqlClient?: Client;
-}
-
-interface WithUrqlInitialProps {
-  urqlState: SSRData;
-}
-
-interface PageProps {
-  pageProps?: WithUrqlClient;
-}
-
-export interface NextUrqlContext extends NextPageContext {
-  urqlClient: Client;
+import {
+  NextUrqlClientConfig,
+  MergeExchanges,
+  NextUrqlContext,
+  WithUrqlProps,
+} from './types';
+
+function getDisplayName(Component: React.ComponentType<any>) {
+  return Component.displayName || Component.name || 'Component';
 }
 
-type NextUrqlClientConfig =
-  | NextUrqlClientOptions
-  | ((ctx?: NextPageContext) => NextUrqlClientOptions);
-
-function withUrqlClient<T = any, IP = any>(
+export function withUrqlClient(
   clientConfig: NextUrqlClientConfig,
-  mergeExchanges: (ssrEx: SSRExchange) => Exchange[] = ssrEx => [
+  mergeExchanges: MergeExchanges = ssrExchange => [
     dedupExchange,
     cacheExchange,
-    ssrEx,
+    ssrExchange,
     fetchExchange,
   ],
 ) {
-  return (Page: NextPage<T & IP & WithUrqlClient, IP>) => {
-    const withUrql: NextComponentType<
-      NextUrqlContext,
-      IP | (IP & WithUrqlInitialProps),
-      T & IP & WithUrqlClient & WithUrqlInitialProps & PageProps
-    > = ({ urqlClient, urqlState, pageProps, ...rest }) => {
-      // The React Hooks ESLint plugin will not interpret withUrql as a React component
-      // due to the NextComponentType annotation. Ignore the warning about not using useMemo.
-
+  return (AppOrPage: NextPage<any> | typeof NextApp) => {
+    const withUrql = ({ urqlClient, urqlState, ...rest }: WithUrqlProps) => {
       // eslint-disable-next-line react-hooks/rules-of-hooks
       const client = React.useMemo(() => {
         const clientOptions =
           typeof clientConfig === 'function' ? clientConfig() : clientConfig;
 
         return (
           urqlClient ||
-          pageProps?.urqlClient ||
-          initUrqlClient(clientOptions, mergeExchanges, urqlState)[0]
+          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+          initUrqlClient(clientOptions, mergeExchanges, urqlState)[0]!
         );
-      }, [urqlClient, pageProps, urqlState]) as Client;
+      }, [urqlClient, urqlState]);
 
       return (
         <Provider value={client}>
-          <Page urqlClient={client} {...(rest as T & IP)} />
+          <AppOrPage urqlClient={client} {...rest} />
```

Comment:
I went for naming this `AppOrPage` – a much better description of what this can be!